# **Smart-UJ UAQUE**
Este Notebook hace parte del servicio inteligente "perfil grupal" y tiene como objetivo la generación de clusters de usuarios a partir de el alquiler de ítems de la biblioteca Alfonso Borrero Cabal. Previamente se estrajeron los temas de cada ítem y en este notebook se generarán matrices dispersas que serán consumidad por el algoritmo de k-means para generar los clusters que posteriormente serán consumidos por los diferentes algoritmo de recomendación.

**Entrada**: Tabla_join.json

**Integrantes:**</br>
Johan Ortegón </br>
Juan Angarita </br>
Jorge Salgado </br>
Daniel Tibaquira </br></br>
**Directora de Proyecto de Grado:** </br>
Viky Arias </br>

# Importación de datos

In [3]:
#from google.colab import drive
#drive.mount('/content/drive')

Estas son las librerias que vamos a usar para hacer el join de ambas tablas y poder hacer las diferentes agrupaciones de estudiantes. 

* pandas(pd): Permite almacenar los datos en dataframes(matrices) y operación entre las celdas.
* KMeans: librería de sklearn para realizar el algoritmo de agrupamiento Kmeans.
* StandartScaler: Permite estandarizar los datos en una escala predeterminada.
* silhoute_score: permite sacar la métrica de puntaje de silhoute para los algoritmos de clustering.
* Normalizer: Permite normalizar los datos
* PCA: principal component analisis permite la reducción de dimensionalidad de nuestros dataset, en busca de que los algoritmos se ejecuten a mayor velocidad.

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import Normalizer
from sklearn.decomposition import PCA

Importamos el archivo. Este proviene del join entre las tablas de material bibliográfico de la biblioteca y la información de prestamos de la biblioteca. 

In [5]:
#df = pd.read_json('https://www.dropbox.com/s/chk4h90xgm48adm/joinTablas.json?dl=1')
df = pd.read_json('https://www.dropbox.com/s/c80i2zi3ba4a1lv/joinTablas-FEEDBACK.json?dl=1')

In [6]:
df.head(3)

level_0  index RowID         Fecha          IDItem NumeroUbicacion Dewey  \
0        0      0  Row0  1.547050e+12  80000005327627     720.98 A71S   720   
1        1      1  Row1  1.547050e+12  80000001191496    720.9861 A71   720   
2        2      2  Row2  1.547139e+12  80000004979759  540 CH15Q 2010   540   

    Ubicacion   Llave                 Programa  ... Calificación Signatura  \
0  COL-GRAL-2  866245  CARRERA DE ARQUITECTURA  ...          NaN       720   
1  COL-GRAL-2  309945  CARRERA DE ARQUITECTURA  ...          NaN       720   
2  COL-GRAL-3  822727  CARRERA DE ARQUITECTURA  ...          NaN       540   

   TipoItem                          Autor  \
0     LIBRO  Arango Cardinal, Silvia 1948-   
1     LIBRO  Arango Cardinal, Silvia 1948-   
2     LIBRO                 Chang, Raymond   

                                              Titulo AnioPublicacion  \
0  Ciudad y arquitectura seis generaciones que co...          2012.0   
1  Historia de la arquitectura en Colombia Silvia...          1993.0   
2  Química Raymond Chang ; revisión técnica Rosa ...          2010.0   

  DeweyEspecifico                     TemaDewey  \
0          720.98  Arquitectura latinoamericana   
1        720.9861       Arquitectura colombiana   
2             540                       Química   

                                       Temas  \
0        [arquitectur, histori, amer, latin]   
1  [arquitectur, colonial, histori, colombi]   
2                                   [quimic]   

                                  Union  
0        arquitectur histori amer latin  
1  arquitectur colonial histori colombi  
2                                quimic  

[3 rows x 23 columns]

Anotamos la estructura del archivo. 

In [7]:
print("Numero de filas:", df.shape[0])
print("Numero de columnas:", df.shape[1])

Numero de filas: 489575
Numero de columnas: 23


El clustering para este experimento se realizará sobre el atributo "Signatura" el cual representa el dewey hasta la unidad (dando aproximadamente 900 deweys diferentes). Recordar la diferencia entre la columna "Dewey" y "Signatura". La primera es la que originalmente contiene el dataset de prestamos y solo venía con valores hasta la decena, mientras que "Sginatura" originalmente era una valor que señalaba el dewey especifico del libro y su ubicación al interior de la biblioteca. Se realizo una transformación para obtener únicamente el valor númerico de este valor y así obtener el dewey hasta la unidad(más especifico).

In [8]:
eliminar = df.loc[df["Signatura"].isnull()]
print("Numero de datos sin Dewey(Singatura):" , eliminar.shape[0])

Numero de datos sin Dewey(Singatura): 0


In [9]:
print("Valores de signatura unicos: " , len(df["Signatura"].unique()))
print("Valores de Dewey(decena) unicos: " , len(df["Dewey"].unique()))

Valores de signatura unicos:  756
Valores de Dewey(decena) unicos:  100


A partir del análisis con expertos en negocio, se ha entendido que es importante tomar en cuenta el cambio de gustos de los usuarios a través del tiempo y dar menos peso a prestamos que se realizaron en el pasado a prestamos más recientes. Para esto se creará una nueva columna denominada "Peso" la cual determinará el peso quue tiene dicho prestamo.
La disminución del peso será exponencial según la diferencia con el año actual y se calculará con la siguiente formula:

$Peso = \frac{1}{2^{(2021-20xx)}}$


In [10]:
def pesos(x):
    peso = 1/2**(2021-x.Year) 
    if x["Calificación"] == -1:
        peso = peso*-1
    return peso

In [11]:
df["Peso"] = df.apply(lambda row: pesos(row) , axis=1 )

In [12]:
df.loc[df.Peso < 0]

level_0  index RowID  Fecha          IDItem NumeroUbicacion Dewey  \
489737   489737      0  None    NaN  80000004921734            None  None   
489738   489738      1  None    NaN  80000003200204            None  None   
489739   489739      2  None    NaN  80000004033284            None  None   
489742   489742      5  None    NaN  80000005573117            None  None   
489747   489747     10  None    NaN  80000003390692            None  None   
489750   489750     13  None    NaN  80000001448045            None  None   
489751   489751     14  None    NaN  80000004453556            None  None   
489752   489752     15  None    NaN  80000003592867            None  None   
489753   489753     16  None    NaN  80000005575202            None  None   
489757   489757     20  None    NaN  80000002477654            None  None   
489758   489758     21  None    NaN  80000005303102            None  None   
489761   489761     24  None    NaN  80000003543500            None  None   
489763   489763     26  None    NaN  80000004221520            None  None   
489766   489766     29  None    NaN  80000005535411            None  None   
489767   489767     30  None    NaN  80000003854491            None  None   
489768   489768     31  None    NaN  80000001233769            None  None   
489771   489771     34  None    NaN  80000002843202            None  None   
489772   489772     35  None    NaN  80000005017344            None  None   
489774   489774     37  None    NaN  80000003302954            None  None   
489775   489775     38  None    NaN  80000001194821            None  None   
489776   489776     39  None    NaN  80000002420142            None  None   
489778   489778     41  None    NaN  80000004110909            None  None   
489783   489783     46  None    NaN  80000002907007            None  None   
489785   489785     48  None    NaN  80000001480220            None  None   
489786   489786     49  None    NaN  80000001428807            None  None   
489787   489787     50  None    NaN  80000001589574            None  None   
489789   489789     52  None    NaN  80000004803566            None  None   
489794   489794     57  None    NaN  80000001932873            None  None   
489796   489796     59  None    NaN  80000003768859            None  None   
489800   489800     63  None    NaN  80000001257297            None  None   
489803   489803     66  None    NaN  80000005056992            None  None   
489806   489806     69  None    NaN  80000001551376            None  None   
489807   489807     70  None    NaN  80000005587598            None  None   
489808   489808     71  None    NaN  80000004104113            None  None   
489809   489809     72  None    NaN  80000001225518            None  None   
489810   489810     73  None    NaN  80000005578741            None  None   
489813   489813     76  None    NaN  80000005094413            None  None   
489818   489818     81  None    NaN  80000004902627            None  None   
489819   489819     82  None    NaN  80000004960511            None  None   
489821   489821     84  None    NaN  80000002134081            None  None   
489824   489824     87  None    NaN  80000005344457            None  None   
489827   489827     90  None    NaN  80000003818264            None  None   
489828   489828     91  None    NaN  80000002439035            None  None   
489832   489832     95  None    NaN  60000000002939            None  None   
489836   489836     99  None    NaN  80000005582653            None  None   
489837   489837    100  None    NaN  80000005327627            None  None   

       Ubicacion Llave Programa  ... Signatura    TipoItem  \
489737      None  None     None  ...       780       LIBRO   
489738      None  None     None  ...       808       LIBRO   
489739      None  None     None  ...       711       LIBRO   
489742      None  None     None  ...       371       LIBRO   
489747      None  None     None  ...       752       LIBRO   
489750      None

Se creará una nueva tabla donde únicamente se van a tener en cuenta los deweys que le gustan a cada usuario así como los atributos para posteriormente realizar el clustering. 

In [13]:
def positivo(x):
    if x<0:
        return 0
    else:
        return x

In [ ]:
pesos_usuarios = pd.DataFrame(columns= df["Signatura"].unique(),dtype=float)
tam = len(df["Signatura"].unique())
pesos_usuarios.columns = pesos_usuarios.columns.astype(str)
display(pesos_usuarios)

ids = df["IDUsuario"].unique()
i = 0
for usuario in ids:
  pesos_usuarios.loc[i] = np.zeros((tam,), dtype=int)
  #Para cada usuario traer sus Deweys y sus pesas asociados
  prestamos = df.loc[df["IDUsuario"]==usuario]
  deweys = prestamos[["Signatura","Peso"]]
  result_user = deweys.groupby("Signatura")["Peso"].sum().reset_index(name="Peso")
  result_user.Peso = result_user.Peso.apply(lambda x: positivo(x))
  # Recorrer los Deweys y Pesos para plasmarlos en una matriz
  #display(result_user)
  for index, row in result_user.iterrows():
    d = row["Signatura"]
    p=row['Peso']
    pesos_usuarios[d][i] = p
  i = i+1

Empty DataFrame
Columns: [720, 540, 512, 712, 741, 814, 823, 330, 863, 696, 620, 744, 158, 943, 728, 574, 724, 711, 843, 745, 809, 307, 690, 301, 333, R, 352, 388, 131, 303, 302, 305, 342, 986, 861, 725, 344, 701, 808, 363, 692, 726, 927, 759, 895, 813, 428, 746, 658, 332, 004, 918, 686, 304, 747, 153, 121, 862, 370, 617, 136, 629, 001, 517, 926, 721, 372, 612, 306, 155, 455, 947, 923, 616, 346, 864, 150, 154, 791, 111, 821, 709, 704, 688, 005, 891, 841, 736, 727, 152, 580, 822, 523, 591, 937, 691, 853, 811, 509, 339, ...]
Index: []

[0 rows x 756 columns]

La matriz generada es una matriz dispersa donde todos los valores son ceros excepto donde el usuario haya realizado un prestamo.

In [ ]:
pesos_usuarios

# Configuración Modelos
A partir del archivo "Comparación modelos" se puede concluir que el modelo que ayuda en mayor medida a agrupar los datos, minimizando el número de clusters con tamaño menor a 20 y la desviación estandar es kmeans con los siguiente parámetros:
* Filas normalizadas = true
* Columnas normalizadas = true
* PCA = 0.95
* Número de clusters = 150
* Número de iteraciones = 500

In [ ]:
resumen = pd.DataFrame(columns = ["Tecnica","PCA", "Norm_filas", "Norm_col", "Precision", "Desviacion_estandar", "Media", "Coeficiente_variacion", "Clusters","Grupo_mas_grande", "Grupos<10", "Grupos<20"])


In [ ]:
def std_dev_cluster(data, labs):
    sizes_labels = []
    for label in np.unique(labs):
        sizes_labels.append(len(data.loc[data["cluster"] == label]))
    
    return np.std(sizes_labels)

In [ ]:
def mean_cluster(data, labs):
    sizes_labels = []
    for label in np.unique(labs):
        sizes_labels.append(len(data.loc[data["cluster"] == label]))
    
    return np.mean(sizes_labels)

In [ ]:
#
PCA = 0.95
num_clusters = 160
num_iter = 1000

# Algoritmo kmeans

In [ ]:
final = pesos_usuarios

In [ ]:
##Exportamos los datos del modelo que nos dio mejores resultados
#Normalización
#2.5 min por iteración
sumatoria = final.sum(axis=1)
pesos_norm = final.div(sumatoria, axis=0).fillna(0)

scaler = MinMaxScaler()
pesos_norm = scaler.fit_transform(pesos_norm.values)

  #Kmeans
k_means = KMeans(init = "k-means++", n_clusters = num_clusters, n_init = num_iter)
k_means.fit_predict(pesos_norm)
labels = k_means.labels_

In [ ]:
#Desempeño
#3 min
pres = silhouette_score(pesos_norm, labels)
print("silhouette_score: ", pres)

In [ ]:
grouped = df.groupby("IDUsuario")["Peso"].sum().reset_index(name="Peso")
grouped

In [ ]:
#copiamos el dataframe y agregamos los clusters para gráficar la distribución
df2 = final.copy()
df2["cluster"] = labels
df2.groupby("cluster")["cluster"].count()
display(df2["cluster"].value_counts().plot(kind='bar'))

#pegamos los cluster a la tabla que tiene los usuarios y los temas
grouped["cluster"] = labels


#Cálculo de métricas
media = mean_cluster(df2,labels)
desviacion = std_dev_cluster(df2,labels)
data = df2.groupby('cluster')['cluster'].count()
peque10 = len(data[data<=10])
peque20 = len(data[data<=20])
masGrande = data.max()
#Agregamos los datos a nuestra tabla de resumen para comparar
resumen = resumen.append({'Tecnica':"K-MEANS",
                              'PCA':0.95,
                              'Norm_filas':True,
                              'Norm_col':True,
                              'Precision':pres,
                              'Desviacion_estandar':desviacion,
                              'Media':media,
                              'Coeficiente_variacion':desviacion/media, 
                              'Clusters': num_clusters,
                              "Grupos<20":peque20,
                              "Grupos<10":peque10 , 
                              "Grupo_mas_grande":masGrande}, ignore_index=True)

In [ ]:
grouped.head(3)

In [ ]:
resumen

Podemos analizar los temas más importantes que tiene cada cluster. 

In [ ]:
cluster = df2.loc[df2.cluster == 89]
aux=cluster.astype(bool).sum(axis=0)
df3 = aux[aux > 0]
df3.sort_values(ascending=False).head(20)

In [ ]:
grouped.head(3)

Pegamos la información del cluster a cada uno de los prestamos de los usuarios.(temas_usuarios_cluster)

In [ ]:
material_data = pd.DataFrame(data=df)
cluster_data = pd.DataFrame(data=grouped)
temas_usuarios_cluster = pd.merge(material_data, cluster_data, left_on='IDUsuario', right_on='IDUsuario', how='left')
temas_usuarios_cluster.head(3)

Crearemos un tercer dataframe el cual contega únicamente la información de cada cluster(sus centroides). 

In [ ]:
centroids = pd.DataFgrame(data = k_means.cluster_centers_)

In [ ]:
centroids.columns = df["Signatura"].unique()
centroids

In [ ]:
centroids[centroids<1e-5] = 0

Finalmente exportamos las tablas para ser consumidas por los sistemas de recomedación junto al dashboard.
* Grouped_temas = usuarios + cluster
* temas_usuarios_cluster = joinTablas + cluster
* Centroids_deweys = clusters + centroides

In [ ]:
grouped.to_json('../data/GruposDeweyUnidad/userCluster.json')

In [ ]:
temas_usuarios_cluster.to_json('../data/GruposDeweyUnidad/joinCluster.json')

In [ ]:
centroids.to_json('../data/GruposDeweyUnidad/centroids.json')